Add sparse semantics context manager.

pearu opened the pull request:
The feature and its usage are described in https://github.com/pytorch/pytorch/pull/94728/files#diff-6d0bd55f94dcbd4ccf83f4006976fa4cb87eb06b72dc16238a21980226be6315R491

This PR is a piece of addressing the problem of interpreting unspecified elements in sparse tensors for autograd.

Historically, functions and their backward functions were written for masked semantics, but now, we target supporting sparse semantics while the masked semantics should be covered by `MaskedTensor` feature. In the meantime, an attempt was made to keep the operations with masked semantics in `torch.sparse` namespace but a clean separation has never been achieved (e.g. the backward of `torch.to_dense` is implemented using masked semantics) and with `MaskedTensor`, `torch.sparse` would not be needed when its implementations are moved under `MaskedTensor`.

This PR introduces a context manager that allows choosing the semantics of operations with sparse tensors, both in Python and Aten, using PyTorch Context. Until we have not clearly separated the default sparse (non-masked) semantics (within `torch` namespace) and masked semantics in `MaskedTensor`, the introduced context manager provides a method of separating the non-masked and masked semantics for autograd methods like `backward` (I presume we don't want to introduce `masked` kw argument to `backward` similar to `torch.autograd.gradcheck`) as well as for backward implementations of various operations (e.g. see the `to_dense_backward` implementation in this PR).

In fact, the suggested feature in this PR will promote moving `torch.sparse` implementations to `torch` implementations on sparse tensors (the implementations can use `at::globalContext().sparseSemantics()` for the expected behavior, `MaskedTensor` operations would disable sparse semantics, for instance).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94728
* #95372



cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer @ezyang @albanD @zou3519 @gqchen @soulitzer @Lezcano @Varal7

pearu committed:
Add sparse semantics context manager.

[ghstack-poisoned]

ezyang commented:
Is there a design writeup about this anywhere?

cpuhrsch commented:
@ezyang - Not from our end. We could write something up if needed. I expect this to require plenty of discussion.

cpuhrsch commented:
@pearu - Could you add a description to this PR?

ezyang commented:
It looks like the info I was looking for is at https://github.com/pytorch/pytorch/pull/94728/files#diff-6d0bd55f94dcbd4ccf83f4006976fa4cb87eb06b72dc16238a21980226be6315R491

amjames commented:
What purpose does this serve beyond what `MaskedTensor` would already provide? With `MaskedTensor` we have a different type to intercept dispatch and apply what is necessary for the masked semantics. With this many ops would need to have masked semantics implemented behind a switch. 

Also the use of "sparse semantics" here I think further confuses the point of what sparsity actually is.

The first PR on this stack #94714 makes sense because sparse support in gradcheck was added to support some ops in the `torch.sparse` namespace which implement masked semantics, and now as we add sparse support for normal operators we need to be able to tell gradcheck that masked gradient semantics is not applicable everywhere that sparse layouts are encountered. 

It has been a long term plan to deprecate those operators from `torch.sparse` which implement masked semantics in favor of `MaskedTensor` as it collects enough features to replace those APIs. This PR seems like a step in the wrong direction.

pearu committed:
Update on "Add sparse semantics context manager."




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add sparse semantics context manager."




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add sparse semantics context manager."




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu commented:
> What purpose does this serve beyond what `MaskedTensor` would already provide? With `MaskedTensor` we have a different type to intercept dispatch and apply what is necessary for the masked semantics. With this many ops would need to have masked semantics implemented behind a switch.

I'd consider it as an intermediate step in cleaning up existing masked-semantics-specific backward implementations without changing the API (e.g. we don't want to introduce `masked` kw arguments to methods like `backward` that must have semantics-dependent behavior).

> Also the use of "sparse semantics" here I think further confuses the point of what sparsity actually is.

I agree, perhaps this should be called "masked semantics", that is, provide `torch.sparse.masked_semantics` context manager to force the current default semantics while the PR would make non-masked semantics the default (BC warning, but we agreed in the meet that this is acceptable BC breaking change).

> The first PR on this stack #94714 makes sense because sparse support in gradcheck was added to support some ops in the `torch.sparse` namespace which implement masked semantics

This is incorrect. The PR #94714 adds sparse tensor support to gradcheck using non-masked semantics. Before, gradcheck was basically broken for non-masked semantics but worked correctly within masked semantics.

> and now as we add sparse support for normal operators we need to be able to tell gradcheck that masked gradient semantics is not applicable everywhere that sparse layouts are encountered.

yes, and more importantly, we'll need to tell backward operations to use the correct semantics. The `masked=False` kw was necessary for correctly computing the numerical gradients but the context manager in this PR is necessary for correctly computing the analytical gradients.

> It has been a long term plan to deprecate those operators from `torch.sparse` which implement masked semantics in favor of `MaskedTensor` as it collects enough features to replace those APIs.

Correct.

> This PR seems like a step in the wrong direction.

Please consider the newly added description of the PR and arguments above.

The main function of the suggested semantics context manager is to enable backward functions to operate within the expected semantics. Currently, this is masked semantics but we want to turn it into sparse semantics, and `at::globalContext().sparseSemantics()` will provide a tool to do that gradually and in BC manner (except for making non-masked semantics the default).

pearu commented:
Btw, with sparse semantics context manager, the `masked` kw argument in `gradcheck` becomes redundant.

albanD commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -504,17 +504,29 @@ Tensor to_dense_backward(const Tensor& grad, const Tensor& input_) {
-      // Autograd operates on the coalesced assumption, i.e. no duplicate values.
-      return grad.sparse_mask(input_.coalesce());
+      if (at::globalContext().sparseSemantics()) {
+        return grad.to_sparse(input_.sparse_dim());
```
I would agree that a no-op is best here. We don't need to try and enforce sparse/sparse correspondance between tensor and gradient.

pearu replied:
I agree. However, I think we are not there yet to enable this for sparse tensors in general.

For instance, when returning `grad`, this will lead to
```
test/test_autograd.py::TestAutograd::test_gradcheck_backward_mul_by_grad_output FAILED
...
_________________________________________________________________________
RuntimeError: hook 'hook' has changed the type of value (was torch.DoubleTensor got torch.sparse.DoubleTensor)
```

amjames replied:
The test you reference should probably be removed. 

Also the checks internal to gradcheck that enforce layout equality between inputs and their gradients on the backward path should go as well. 

In general backward implementations can make no assumption about the layout of the incoming grad and should not do anything to artificially impose a layout on the gradients coming out.

pearu replied:
> The test you reference should probably be removed.

Possibly. I don't know the extent of this issue yet to decide if this suggestion should be implemented in this PR or in a follow-up, especially when the feature requires "plenty of discussions".

albanD reviewed (commented):
That's an interesting approach.
Is this limited to backward? Or do we want to update optimizers / softmax to also follow that?

amjames commented:
These questions by @albanD get right to the point of a major reservation I have about this:

> Is this limited to backward? Or do we want to update optimizers / softmax to also follow that?

If this context manager exists it may lead to assumptions by users about what it can do. This PR introduces semantics based behavior for `to_dense_backward`. Is the plan to then add a switch to all backward functions?  Then later on remove all that as we add functionality to `MaskedTensor`? What about the forward functions?

pearu commented:
> These questions by @albanD get right to the point of a major reservation I have about this:

Can you suggest alternative approaches to the problem? To reiterate, the problem is about adopting existing operations (that currently implement masked semantics support) to support non-masked semantics in the least BC-breaking manner.

> > Is this limited to backward? Or do we want to update optimizers / softmax to also follow that?

The answer depends on if these need to support both masked and non-masked semantics, or not.

> If this context manager exists it may lead to assumptions by users about what it can do. This PR introduces semantics based behavior for `to_dense_backward`.

It is not limited to `to_dense_backward`, it is just the simplest example that currently supports masked semantics (for historical reasons) and this PR adds support to the non-masked semantics (which we want) without breaking existing `torch.sparse` related tests.

> Is the plan to then add a switch to all backward functions?

Why? If the behavior of masked and non-masked semantics in a backward function is identical, then there is no need to add the switch.

> Then later on remove all that as we add functionality to `MaskedTensor`?

It depends, not necessarily:
`MaskedTensor` is a different construction (it is a pair of values and mask tensors) from sparse tensors under masked semantics. The latter is a subset of `MaskedTensor` (operations on tensors with identical sparsity patterns) that is more efficient (no need for the mask tensors as the sparsity pattern defines it). So, a sparse tensor under masked semantics can be considered as an optimization of `MaskedTensor`. This is similar to a sparse tensor being an optimization of a strided tensor under non-masked semantics. So, I would not consider sparse tensors under masked semantics as a bad thing in terms of performance.

> What about the forward functions?

The answer is the same as above. If the behavior of forward functions on sparse tensors depends on the semantics, there should be a switch.

albanD commented:
Thanks for the clarification @pearu 
I am not sure how the context manager based policy will work in practice but I think it is worth doing it to unblock users and clarify what we are doing.
I do agree that it should cover all PyTorch functions that might have different semantic for sparse vs masked.

One question I would have is about the default state? Are we planning on allowing a third state where each op chooses what it does (like softmax vs add) or we're planning to make the appropriate BC-breaking changes to move everything to the semantic from the flag?

pearu commented:
> One question I would have is about the default state? Are we planning on allowing a third state where each op chooses what it does (like softmax vs add) or we're planning to make the appropriate BC-breaking changes to move everything to the semantic from the flag?

IIUC, the third state case suggestion would actually involve two extra states on top of the masked and non-masked semantics, that is, the four states would be:
- all operations use masked semantics
- all operations use non-masked semantics
- certain operations use their preferred semantics (masked or non-masked) but all other operations use masked semantics
- certain operations use their preferred semantics (masked or non-masked) but all other operations use non-masked semantics

These extra states can be avoided because the semantics context managers can be nested. For instance:
```python
# default semantics is non-masked

# the default can be redefined via
torch.sparse.sparse_semantics.disable()  # default will be masked semantics
# or
torch.sparse.sparse_semantics.enable()   # default will be non-masked semantics

with torch.sparse.sparse_semantics(False):
  x = torch.softmax(...)  # it will use masked semantics
  with torch.sparse.sparse_semantics(True):
     ...  # ops here use non-masked semantics
  ... # back to using masked semantics
... # back to using the default semantics
```
The autograd graph will certainly look interesting when nesting semantics in this way: it will contain backwards of the same operations but these backwards may use different semantics on different leaves. @albanD , would this be problematic to autograd in any way?

Regarding the naming conventions:
1. I'd suggest replacing `torch.sparse.sparse_semantics(False)` with `torch.sparse.masked_semantics()` so that there will be two semantics context managers (`sparse_semantics` and `masked_semantics`) that manage the same Context flag.
2. Or, another option is to use `torch.sparse.semantics("masked")` or `torch.sparse.semantics("non-masked")` as the user interface.
What do you think?

pearu committed:
Update on "Add sparse semantics context manager."

The feature and its usage are described in https://github.com/pytorch/pytorch/pull/94728/files#diff-6d0bd55f94dcbd4ccf83f4006976fa4cb87eb06b72dc16238a21980226be6315R491

This PR is a piece of addressing the problem of interpreting unspecified elements in sparse tensors for autograd.

Historically, functions and their backward functions were written for masked semantics, but now, we target supporting sparse semantics while the masked semantics should be covered by `MaskedTensor` feature. In the meantime, an attempt was made to keep the operations with masked semantics in `torch.sparse` namespace but a clean separation has never been achieved (e.g. the backward of `torch.to_dense` is implemented using masked semantics) and with `MaskedTensor`, `torch.sparse` would not be needed when its implementations are moved under `MaskedTensor`.

This PR introduces a context manager that allows choosing the semantics of operations with sparse tensors, both in Python and Aten, using PyTorch Context. Until we have not clearly separated the default sparse (non-masked) semantics (within `torch` namespace) and masked semantics in `MaskedTensor`, the introduced context manager provides a method of separating the non-masked and masked semantics for autograd methods like `backward` (I presume we don't want to introduce `masked` kw argument to `backward` similar to `torch.autograd.gradcheck`) as well as for backward implementations of various operations (e.g. see the `to_dense_backward` implementation in this PR).

In fact, the suggested feature in this PR will promote moving `torch.sparse` implementations to `torch` implementations on sparse tensors (the implementations can use `at::globalContext().sparseSemantics()` for the expected behavior, `MaskedTensor` operations would disable sparse semantics, for instance).




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

Skylion007 commented on torch/sparse/__init__.py:
```diff
@@ -485,3 +486,113 @@ def test_mth(*args, **kwargs):
+    # context manager support
+    def __init__(self, enable=True):
+        self.state = enable
+        self.saved_state = self.is_enabled()
```
Technically, shoudn't self.saved_state be cached on ```__enter__``` and not on __init__. You could imagine a rare cause where the context manager is initiated way before it actually reaches the portion of the code where it's called using a with statement.

pearu replied:
Yes, thanks for pointing this out! In fact, the bug wouldn't be so rare. Here's a reproducer:
```python
assert torch.sparse.sparse_semantics.is_enabled() == True
masked_ctx = torch.sparse.sparse_semantics(False)
non_masked_ctx = torch.sparse.sparse_semantics(True)
with masked_ctx:
    assert torch.sparse.sparse_semantics.is_enabled() == False
    with non_masked_ctx:
        assert torch.sparse.sparse_semantics.is_enabled() == True
    assert torch.sparse.sparse_semantics.is_enabled() == False  # FAILURE
```
I'll fix it.

pearu replied:
Fixed.

pearu committed:
Update on "Add sparse semantics context manager."

The feature and its usage are described in https://github.com/pytorch/pytorch/pull/94728/files#diff-6d0bd55f94dcbd4ccf83f4006976fa4cb87eb06b72dc16238a21980226be6315R491

This PR is a piece of addressing the problem of interpreting unspecified elements in sparse tensors for autograd.

Historically, functions and their backward functions were written for masked semantics, but now, we target supporting sparse semantics while the masked semantics should be covered by `MaskedTensor` feature. In the meantime, an attempt was made to keep the operations with masked semantics in `torch.sparse` namespace but a clean separation has never been achieved (e.g. the backward of `torch.to_dense` is implemented using masked semantics) and with `MaskedTensor`, `torch.sparse` would not be needed when its implementations are moved under `MaskedTensor`.

This PR introduces a context manager that allows choosing the semantics of operations with sparse tensors, both in Python and Aten, using PyTorch Context. Until we have not clearly separated the default sparse (non-masked) semantics (within `torch` namespace) and masked semantics in `MaskedTensor`, the introduced context manager provides a method of separating the non-masked and masked semantics for autograd methods like `backward` (I presume we don't want to introduce `masked` kw argument to `backward` similar to `torch.autograd.gradcheck`) as well as for backward implementations of various operations (e.g. see the `to_dense_backward` implementation in this PR).

In fact, the suggested feature in this PR will promote moving `torch.sparse` implementations to `torch` implementations on sparse tensors (the implementations can use `at::globalContext().sparseSemantics()` for the expected behavior, `MaskedTensor` operations would disable sparse semantics, for instance).




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

Skylion007 commented on torch/sparse/__init__.py:
```diff
@@ -485,3 +486,116 @@ def test_mth(*args, **kwargs):
+    # context manager support
+    def __init__(self, enable=True):
+        self.state = enable
+        self.saved_state = None
```
```suggestion
        self.saved_state : Optional[bool] = None
```
to fix the lintrunner error.

pearu replied:
Applied, thanks!

Skylion007 commented on torch/sparse/__init__.py:
```diff
@@ -485,3 +486,116 @@ def test_mth(*args, **kwargs):
+        torch._C._set_sparse_semantics(self.state)
+
+    def __exit__(self, type, value, traceback):
+        torch._C._set_sparse_semantics(self.saved_state)
```
```suggestion
        assert self.saved_state is not None
        torch._C._set_sparse_semantics(self.saved_state)
```

pearu replied:
Applied, thanks!

pearu committed:
Update on "Add sparse semantics context manager."

The feature and its usage are described in https://github.com/pytorch/pytorch/pull/94728/files#diff-6d0bd55f94dcbd4ccf83f4006976fa4cb87eb06b72dc16238a21980226be6315R491

This PR is a piece of addressing the problem of interpreting unspecified elements in sparse tensors for autograd.

Historically, functions and their backward functions were written for masked semantics, but now, we target supporting sparse semantics while the masked semantics should be covered by `MaskedTensor` feature. In the meantime, an attempt was made to keep the operations with masked semantics in `torch.sparse` namespace but a clean separation has never been achieved (e.g. the backward of `torch.to_dense` is implemented using masked semantics) and with `MaskedTensor`, `torch.sparse` would not be needed when its implementations are moved under `MaskedTensor`.

This PR introduces a context manager that allows choosing the semantics of operations with sparse tensors, both in Python and Aten, using PyTorch Context. Until we have not clearly separated the default sparse (non-masked) semantics (within `torch` namespace) and masked semantics in `MaskedTensor`, the introduced context manager provides a method of separating the non-masked and masked semantics for autograd methods like `backward` (I presume we don't want to introduce `masked` kw argument to `backward` similar to `torch.autograd.gradcheck`) as well as for backward implementations of various operations (e.g. see the `to_dense_backward` implementation in this PR).

In fact, the suggested feature in this PR will promote moving `torch.sparse` implementations to `torch` implementations on sparse tensors (the implementations can use `at::globalContext().sparseSemantics()` for the expected behavior, `MaskedTensor` operations would disable sparse semantics, for instance).




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add sparse semantics context manager."

The feature and its usage are described in https://github.com/pytorch/pytorch/pull/94728/files#diff-6d0bd55f94dcbd4ccf83f4006976fa4cb87eb06b72dc16238a21980226be6315R491

This PR is a piece of addressing the problem of interpreting unspecified elements in sparse tensors for autograd.

Historically, functions and their backward functions were written for masked semantics, but now, we target supporting sparse semantics while the masked semantics should be covered by `MaskedTensor` feature. In the meantime, an attempt was made to keep the operations with masked semantics in `torch.sparse` namespace but a clean separation has never been achieved (e.g. the backward of `torch.to_dense` is implemented using masked semantics) and with `MaskedTensor`, `torch.sparse` would not be needed when its implementations are moved under `MaskedTensor`.

This PR introduces a context manager that allows choosing the semantics of operations with sparse tensors, both in Python and Aten, using PyTorch Context. Until we have not clearly separated the default sparse (non-masked) semantics (within `torch` namespace) and masked semantics in `MaskedTensor`, the introduced context manager provides a method of separating the non-masked and masked semantics for autograd methods like `backward` (I presume we don't want to introduce `masked` kw argument to `backward` similar to `torch.autograd.gradcheck`) as well as for backward implementations of various operations (e.g. see the `to_dense_backward` implementation in this PR).

In fact, the suggested feature in this PR will promote moving `torch.sparse` implementations to `torch` implementations on sparse tensors (the implementations can use `at::globalContext().sparseSemantics()` for the expected behavior, `MaskedTensor` operations would disable sparse semantics, for instance).




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

Skylion007 commented on torch/sparse/__init__.py:
```diff
@@ -491,3 +492,119 @@ def test_mth(*args, **kwargs):
+        torch._C._set_sparse_semantics(self.saved_state)
+        self.saved_state = None
+
+    # decorator support
```
@pearu Wouldn't it just be cleaner to have it inherit from contextlib.ContextDecorator?

pearu commented:
This PR is shelved as a draft because the semantics context manager is not proven to resolve the problem of the expected behavior of gradcheck within masked semantics, see discussions in https://github.com/pytorch/pytorch/pull/95405 and https://github.com/pytorch/pytorch/issues/95550.
